simulator/mips/t/unit_test.cpp

```diff
@@ -243,3 +206,454 @@ TEST_CASE( "MIPS32_instr_disasm: Process_Disasm_Branches")
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0);
+
+    instr.set_v_src( 0x1000, 0);
+    instr.set_shamt( 4);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0x100);
+
+    // Little endian
+    instr.set_v_src( 0xff000000, 0);
+    instr.set_shamt( 8);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0xffff0000);
+}
+
+TEST_CASE( "MIPS32_instr: sllv")
+{
+    MIPS32Instr instr( "sllv");
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 0, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 1);
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 1, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 2);
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 16, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0x10000);
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 32, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 1);
+}
+
+TEST_CASE( "MIPS32_instr: srlv")
+{
+    MIPS32Instr instr( "srlv");
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 0, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 1);
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 1, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0);
+
+    instr.set_v_src( 0x1000, 0);
+    instr.set_v_src( 4, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0x100);
+
+    instr.set_v_src( 0x80000000, 0);
+    instr.set_v_src( 4, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0x08000000);
+}
+
+// Not implemented in SPIM TT
+TEST_CASE( "MIPS32_instr: srav")
+{
+    MIPS32Instr instr( "srav");
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 0, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 1);
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 1, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0);
+
+    instr.set_v_src( 0x1000, 0);
+    instr.set_v_src( 4, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0x100);
+
+    // Little endian
+    instr.set_v_src( 0xff000000, 0);
+    instr.set_v_src( 4, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0xfff00000);
+}
+
+TEST_CASE( "MIPS32_instr: jr")
+{
+    MIPS32Instr instr( "jr");
+
+    instr.set_new_PC( instr.get_PC() + 4);
+    instr.set_v_src( 1024, 0);
+    instr.execute();
+    CHECK( instr.get_new_PC() == 1024);
+
+    instr.set_new_PC( instr.get_PC() + 4);
+    instr.set_v_src( instr.get_PC() + 2, 0);
+    instr.execute();
+    CHECK( instr.get_new_PC() == instr.get_PC() + 4);
+}
+
+TEST_CASE( "MIPS32_instr: jalr")
+{
+    MIPS32Instr instr( "jalr");
+
+    instr.set_new_PC( instr.get_PC() + 4);
+    instr.set_v_src( instr.get_PC() + 8, 0);
+    instr.execute();
+    CHECK( instr.get_new_PC() == instr.get_PC() + 8);
+    CHECK( instr.get_v_dst() == instr.get_PC() + 4);
+
+    instr.set_new_PC( instr.get_PC() + 4);
+    instr.set_v_src( instr.get_PC() + 2, 0);
+    instr.execute();
+    CHECK( instr.get_new_PC() == instr.get_PC() + 4);
+}
+
+/*
+TEST_CASE( "MIPS32_instr: movz")
```

Comment:
movz/movn are implemented on the register file level, so we are not able to test them here.

Reply:
Thanks.